### ISSUE TITLE
Bug: Components inside typescript namespaces cause ReferenceError

### ISSUE THREAD

**Issue by Enteleform** [facebook/react#22413]
Forwarded from https://github.com/vitejs/vite/issues/3900 by @not-rusty based on the recommendation of @sodatea, which stated that this should be considered as a bug in the `react-refresh` package.
&nbsp;  

----

### Describe the bug
Apparently components inside namespaces is not supported. I don't exactly know if the error comes from esbuild or something, but it would nice.

The bug is that is not shown as a compilation error.

### Reproduction
Simply start a `react-ts` project with `$ npm init @vitejs/app my-vue-app --template react-ts` and write the following code:

```ts
namespace Lol {
  export const Lol = () ={
    return (
      <div>
        <p>Some component</p>
      </div>
    );
  };
}

function App() {
  return (
    <div className="App">
      <header className="App-header">
        <Lol.Lol />
        <p>Hello Vite + React!</p>
      </header>
    </div>
  );
}

export default App;
```

### System Info
```js
  System:
    OS: Linux 5.8 Ubuntu 20.04.2 LTS (Focal Fossa)
    CPU: (8) x64 Intel(R) Core(TM) i5-10210U CPU @ 1.60GHz
    Memory: 440.54 MB / 7.47 GB
    Container: Yes
    Shell: 5.8 - /usr/bin/zsh
  Binaries:
    Node: 14.16.0 - /usr/local/bin/node
    npm: 6.14.13 - /usr/local/bin/npm
  Browsers:
    Chrome: 90.0.4430.93
    Firefox: 89.0.1
```

Used package manager: npm

### Logs
I get the following runtime error:

```js
App.tsx:13 Uncaught ReferenceError: _c is not defined
    at App.tsx:13
    at App.tsx:14
```

I suppose it should be a compilation error?

**Comment by irinakk** [comment]
To add more details, namespace is transformed into this piece of code:
```
var N;
(function(N2) {
  N2.Test = () => {
  };
  _c = N2.Test; // Uncaught ReferenceError: _c is not defined
})(N || (N = {}));
```
Thee error occurs in vite's development mode since the application code is imported via `<script type="module">`, which creates strict mode.

**Comment by bvaughn** [comment]
cc @gaearon To confirm if this is a known/expected limitation.

**Comment by todd-elvers** [comment]
I have also encountered this error using Vite & react-refresh on a TypeScript react app that has react components nested under namespaces

**Comment by gaearon** [comment]
Should be fixed in `react-refresh@0.11.0`. Please ask Vite to upgrade and see if that fixes it.
